sdk/metric/aggregator/exponential/structure/exponential.go

```diff
@@ -0,0 +1,657 @@
+// limitations under the License.
+
+package structure // import "go.opentelemetry.io/otel/sdk/metric/aggregator/exponential/structure"
+
+import (
+	"fmt"
+	"math/bits"
+
+	"go.opentelemetry.io/otel/sdk/metric/aggregator/exponential/mapping"
+	"go.opentelemetry.io/otel/sdk/metric/aggregator/exponential/mapping/exponent"
+	"go.opentelemetry.io/otel/sdk/metric/aggregator/exponential/mapping/logarithm"
+)
+
+type (
+	// Histogram observes counts observations in exponentially-spaced
+	// buckets.  It is configured with a maximum scale factor
+	// which determines resolution.  Scale is automatically
+	// adjusted to accommodate the range of input data.
+	//
+	// Note that the generic type `N` determines the type of the
+	// Sum, Min, and Max fields.  Bucket boundaries are handled in
+	// floating point regardless of the type of N.
+	Histogram[N ValueType] struct {
+		// maxSize is the maximum capacity of the positive and
+		// negative ranges.  it is set by Init(), preserved by
+		// Copy and Move.
+		maxSize int32
+
+		// sum is the sum of all Updates reflected in the
+		// aggregator.  It has the same type number as the
+		// corresponding sdkinstrument.Descriptor.
+		sum N
+		// count is incremented by 1 per Update.
+		count uint64
+		// zeroCount is incremented by 1 when the measured
+		// value is exactly 0.
+		zeroCount uint64
+		// min is set when count > 0
+		min N
+		// max is set when count > 0
+		max N
+		// positive holds the positive values
+		positive Buckets
+		// negative holds the negative values in these buckets
+		// by their absolute value.
+		negative Buckets
+		// mapping corresponds to the current scale, is shared
+		// by both positive and negative ranges.
+		mapping mapping.Mapping
+	}
+
+	// Buckets stores counts for measurement values in the range
+	// (0, +Inf).
+	Buckets struct {
+		// backing is a slice of nil, []uint8, []uint16, []uint32, or []uint64
+		backing bucketsBacking
+
+		// indexBase is index of the 0th position in the
+		// backing array, i.e., backing[0] is the count associated with
+		// indexBase which is in [indexStart, indexEnd]
+		indexBase int32
+
+		// indexStart is the smallest index value represented
+		// in the backing array.
+		indexStart int32
+
+		// indexEnd is the largest index value represented in
+		// the backing array.
+		indexEnd int32
+	}
+
+	// ValueType is an interface constraint for the numeric type
+	// aggregated by this histogram.
+	ValueType interface {
+		int64 | float64
+	}
+
+	// bucketsCount are the possible backing array widths.
+	bucketsCount interface {
+		uint8 | uint16 | uint32 | uint64
+	}
+
+	// bucketsVarwidth is a variable-width slice of unsigned int counters.
+	bucketsVarwidth[N bucketsCount] struct {
+		counts []N
+	}
+
+	// bucketsBacking is implemented by bucektsVarwidth[N].
+	bucketsBacking interface {
+		// size returns the physical size of the backing
+		// array, which is >= buckets.Len() the number allocated.
+		size() int32
```

Comment:
should this be uint32?

Reply:
Commented: I believe the code is more readable leaving this as int32. Basically, there's a lot of arithmetic involving int32 that this number features in.